src/colmap/estimators/generalized_pose.h

```diff
@@ -0,0 +1,97 @@
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of ETH Zurich and UNC Chapel Hill nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Author: Johannes L. Schoenberger (jsch-at-demuc-dot-de)
+
+#pragma once
+
+#include "colmap/estimators/pose.h"
+#include "colmap/geometry/rigid3.h"
+#include "colmap/optim/ransac.h"
+#include "colmap/scene/camera.h"
+
+#include <vector>
+
+#include <Eigen/Core>
+
+namespace colmap {
+
+// Estimate generalized absolute pose from 2D-3D correspondences.
+//
+// @param options              RANSAC options.
+// @param points2D             Corresponding 2D points.
+// @param points3D             Corresponding 3D points.
+// @param camera_idxs          Index of the rig camera for each correspondence.
+// @param cams_from_rig        Relative pose from rig to each camera frame.
+// @param cameras              Cameras for which to estimate pose.
+// @param rig_from_world       Estimated rig from world pose.
+// @param num_inliers          Number of inliers in RANSAC.
+// @param inlier_mask          Inlier mask for 2D-3D correspondences.
+//
+// @return                     Whether pose is estimated successfully.
+bool EstimateGeneralizedAbsolutePose(
+    const RANSACOptions& options,
+    const std::vector<Eigen::Vector2d>& points2D,
+    const std::vector<Eigen::Vector3d>& points3D,
+    const std::vector<size_t>& camera_idxs,
+    const std::vector<Rigid3d>& cams_from_rig,
+    const std::vector<Camera>& cameras,
+    Rigid3d* rig_from_world,
+    size_t* num_inliers,
+    std::vector<char>* inlier_mask);
```

Comment:
@ahojnnes  Why are inlier masks defined as `vector<char>` instead of `vector<bool>`?

Reply:
vector<bool> uses a bit mask under the hood, so each read or write requires some additional shuffling of bits. I vaguely remember benchmarking this in the past and then using char instead. Not sure if this applies here though. I would keep it as is, unless we have opposite evidence from some profiling.